### PR TITLE
Enforce `Name` not being a valid `UUID`

### PR DESCRIPTION
**Pull request by zephraph**
Back before the `/by-id/` approach was decided on I was trying to make ids and names interchangeable. One of the first things I ran across was the need to ensure that the `Name` type couldn't be a valid `UUID` so we could parse it reliably. We've decided not to take that tack but as I was reflecting I still believe it'd be a good idea to enforce this separation.

First off, not all UUIDs are a valid `Name` today. A `Name` must start with a lowercase letter which means any `UUID` starting with a number is invalid. If a user adds a valid `Name` which also happens to be a valid `UUID` and that shows up in the UI it could be potentially confusing. By adding this restriction users could still use UUIDs as a name when combined with a prefix or suffix, especially given that UUIDs are 36 characters in length and `Name` supports up to a maximum of 63. 

Further this gives us potential future product optionality. It is possible that the console itself could accept `Name` or `UUID` as apart of its path parameters to [enable the linking of resources without requiring a lookup to build the URL](https://github.com/oxidecomputer/console/pull/1047#issuecomment-1183592346). In that case the lookup would only need to happen once on navigation and not for every rendered link. 

While the above functionality isn't a given, the cost of this restriction is negligible, it does avoid confusion in the UI, and it preserves our options for similar future improvements.

### CODE REVIEW

**Review by ahl — COMMENTED** [review]
I like this future-proofing / anti-foot-gunning measure.

I can't comment on the `impl JsonSchema for Name` because github won't let me, but I think it's worth considering what (if anything) we should do there. I see 3 options:

1. Do nothing (always popular)
2. Add some text to the description "(a name cannot be interpretable as a UUID)"
3. use fancy JSON schema to represent the incompatibility

I think 2 probably makes the most sense. 3 is probably overly cute and would make the schema much more complicated to understand. I think it would look something like this:

```json
{
  "allOf": [
    {
      "type": "string",
      "pattern": "^[a-z](|[a-z0-9-]*[a-z0-9])$"
    },
    {
      "not": [
        {
          "type": "string",
          "format": "uuid"
        }
      ]
    }
  ]
}
```

(maybe it's not so bad... I don't know... what do you think?)

**Comment by davepacheco** [comment]
Thanks for looking at this.

I wonder if it'd be clearer and cleaner (for users too) if we changed the existing `Name` regex to somehow exclude uuids.  I can't think of a great way to do that though.

This particular approach ("anything that matches this regexp, unless it also matches this other one") feels a little fraught.  For one, it would seem dodgy (to me) for us in the future to accept some text string and guess whether it's a Name or Id based only on whether it parses as an id.  I generally find these heuristics frustrating as a user.  I can see two cases where this could be annoying for users even today:

- They might want to use uuids for names, if they're building automation that's creating things.  Uuids would be a decent choice for names since they're unique and the automation could create and store them before having created the corresponding API resource.  (It's one annoying thing about the server-generated ids that you can't know them before you create the resource, as we saw in the recent work on the unauthorized.rs test.)
- In a similar case, they might randomly generate random names and they could _happen_ to match the uuid regexp.  To work around this problem, their random name generator would have to check if the randomly-generated name happens to parse as a UUID and retry.  That seems not great?

These might be unlikely?  Or maybe I'm overthinking this?

**Comment by zephraph** [comment]
@ahl, I'll update the openapi docs for sure. As far as the special openapi usage... my first reaction is _not_ to do it just given the downstream complexity of generators having to handle that (cc @david-crespo, @karencfv). As a follow up thought though... we have cases where we need to reserve names where we aren't currently. For example, `new` should be a reserved b/c it's a valid route in the console. There will likely be many other reserved names. We don't currently have a way to represent this, but perhaps that _is_ something we'd want to represent in the API spec?

---

> For one, it would seem dodgy (to me) for us in the future to accept some text string and guess whether it's a Name or Id based only on whether it parses as an id

I'd like to understand this more. Is it based on passed experience, specific technical challenges, or personal preference? I've worked on a few systems in the past that allows ids/slugs to be used interchangeably without issues that I'm aware of. This is especially common in the rails world with folks using Gems like [FriendlyID](https://norman.github.io/friendly_id/file.Guide.html) to facilitate this and other features (like historical slug redirects). 

> They might want to use uuids for names, if they're building automation that's creating things

This isn't viable given that `Name` is restricted to starting with a lowercase ascii character and it's valid for a UUID to start with a number. If a user wanted to do this that'd _have_ to add a prefix so it's no different than our current state.

> they might randomly generate random names and they could happen to match the uuid regexp

This is something I also don't think is much of an issue. They'd need to make sure whatever they randomly generated was a valid name anyway. Again, probably the easiest way to do this is to have a static prefix and some random string (even a UUID) appended to it. 

---

I did take a look at seeing if I could update the regex to not match against UUIDs, but I'm not sure we'd be able to do that.

**Comment by zephraph** [comment]
I stand corrected on the regex. I think this'll work:

```js
/^(?!\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b$)[a-z](|[a-zA-Z0-9-]*[a-zA-Z0-9])$/
```

https://regexr.com/6ppbt

image

**Comment by davepacheco** [comment]
> I'd like to understand this more. Is it based on passed experience, specific technical challenges, or personal preference?

I think it's a personal preference based on past experience with surprising behavior in similar-sounding schemes.  I agree it's convenient when it works, and it's especially valuable when it's a human that's providing the input string.  But for machine interfaces, I think there's value in the interface being specific about what's intended.  It feels analogous to strong typing.

It also feels like the sort of thing that could lead to security breaches (because of how easy it seems like it would be to cause somebody to operate on something different than they expected) but I can't really think of how.

>> They might want to use uuids for names, if they're building automation that's creating things

> This isn't viable given that Name is restricted to starting with a lowercase ascii character and it's valid for a UUID to start with a number. If a user wanted to do this that'd have to add a prefix so it's no different than our current state.

Good point!

>> they might randomly generate random names and they could happen to match the uuid regexp

> This is something I also don't think is much of an issue. They'd need to make sure whatever they randomly generated was a valid name anyway. Again, probably the easiest way to do this is to have a static prefix and some random string (even a UUID) appended to it.

🤷 That doesn't strike me as that uncommon, though it probably _is_ unlikely that the random string happens to look like a valid uuid.

**Review by ahl — APPROVED** [review]
looks good to me!

File: nexus/tests/integration_tests/basic.rs [hunk 1]
```diff
@@ -476,10 +476,10 @@ async fn test_projects_list(cptestctx: &ControlPlaneTestContext) {
     for _ in 0..projects_total {
         // We'll use uuids for the names to make sure that works, and that we
         // can paginate through by _name_ even though the names happen to be
-        // uuids.  Names have to start with a letter, though, so we've got to
-        // make sure our uuid has one.
+        // uuids.  Names have to start with a letter and can't exactly match
```
Comment (davepacheco) [inline nexus/tests/integration_tests/basic.rs:479]: I can't believe we actually had a case doing exactly this!  How was it working before?

Comment (zephraph) [reply nexus/tests/integration_tests/basic.rs:479]: Do you mean the test? The first character of the UUID was replaced with a lowercase ascii character so it was a bit artificial.

Comment (davepacheco) [reply nexus/tests/integration_tests/basic.rs:479]: Oh of course 🤦

**Pull request merged**